### ISSUE TITLE
Exclude self-by-value trait methods implemented on Box<T> from boxed_local

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#1478]
Simplest case:

```rust
trait Trait {
    fn f(self);
}

impl<T> Trait for Box<T> {
    fn f(self) {}
}
```

```
warning: local variable doesn't need to be boxed here, #[warn(boxed_local)] on by default
 --> src/main.rs:9:10
  |
9 |     fn f(self) {}
  |          ^^^^
```

More realistic case that we hit in https://github.com/dtolnay/syn/pull/85:

```rust
trait LiftOnce<T, U> {
    type Output;
    fn lift<F>(self, f: F) -> Self::Output where F: FnOnce(T) -> U;
}

impl<T, U> LiftOnce<T, U> for Box<T> {
    type Output = Box<U>;
    fn lift<F>(self, f: F) -> Box<U> where F: FnOnce(T) -> U {
        Box::new(f(*self))
    }
}
```

I would consider this a false positive.

**Comment by sinkuu** [comment]
There is also false-positive with `FnBox::call_box`. Maybe all `self: Box<_>` should be excluded from boxed_local lint.

**Comment by mcarton** [comment]
@oli-obk E-easy? Do we have a method to check if a type implements `Foo<…>` for any `…` now? (here `Box<T>` for any `T`).

**Comment by oli-obk** [comment]
I'd have just excluded all impl trait for Box methods.

**Comment by phansch** [comment]
Fixed via #3321 (feel free to re-open if it's still a problem)

**Comment by phansch** [comment]
Fixed via #3321 (feel free to re-open if it's still a problem)
